docs/design/object-locking.md

```diff
@@ -0,0 +1,187 @@
+
+**Buckets** -  retention period can be set in either `days` or `years`, but not
+both at the same time.
+
+**Objects** - retention period can be set as date and time when the object is
+expected to expire.
+
+### Legal Hold
+
+Legal hold can be enabled on an object version. Once a legal hold is enabled,
+regardless of the object's retention date or retention mode, the object version
+cannot be deleted until the legal hold is removed.
+Legal hold can be set on an object version during PUT Object request by setting
+`x-amz-object-lock-legal-hold` header or using PUT Object Legal Hold api
+request.
+Root users with account credentials or IAM users who are given the permission
+`s3:PutObjectLegalHold` are allowed to set Legal hold on an object version.
+
+### Implementation
+
+#### Storing Object lock configuration
+
+Object lock can be enabled on a bucket during bucket creation using
+`x-amz-bucket-object-lock-enabled` header. This is stored along with the
+bucket's metadata defined in BucketInfo model.
+Any default lock settings (governance/compliance mode, days/years for retention
+period) defined using `PutObjectLockConfiguration` request are stored along
+with the Bucket's metadata.
+When a PUT Object request is received, the lock configuration is evaluated and
+retain until date and time is calculated and set as `retain-until-date`
+property on the object's metadata.
+Lock configuration on a PUT object request is evaluated the following order
+
+1. Bucket's configuration for retention mode/period
+2. Object's configuration in the PUT Object for retention mode/period
+
+The object's settings override the bucket's settings when calculating the
+`retain-until-date` date and time to be stored on the object's metadata.
+Lock configuration on an object version can also be changed using
+`PutObjectRetention` api request, when the object has no prior retention mode
+set or if the object has `GOVERNANCE` mode set and the user has appropriate
+permission to make the request.
+
+Note: Delete markers do not have any object lock protection
+
+#### Processing DELETE requests
+
+Whenever requests such as  DELETE object using version-id,
+Multi-Object DELETE specifying version-id or a Lifecycle action to
+permanently delete the object version are received, the current date and time is
+evaluated against `retain-until-date` set on the object and the client will
+receive `Access Denied` error if the current date and time is less than the
+`retain-until-date` set on the object.
+When the current date and time exceeds the `retain-until-date`, deletes on the
+object version are automatically allowed. There is no cleanup action to remove
+the `retain-until-date` set on the object version's metadata once the
+`retain-until-date` expires.
+DELETE object requests without a version id result in creation of delete markers
+on top of the object version, even if the object has a lock configuration is set
+and the `retain-until-date` is current.
+
```

Comment:
What is the Lifecycle supposed to do when receiving 'Access Denied' ? Shall it ignore ? maintain metrics ?

Reply:
It needs to be tested but I think Lifecycle currently ignores. Maintaining metrics would make sense but I am not sure how LIfecycle can figure out if the Access Denied came because of lock or not.